File: 프로그래머스/1/133499. 옹알이 （2）/옹알이 （2）.cs
using System;
using System.Text.RegularExpressions;

public class Solution 
{
    public int solution(string[] babbling) 
    {
        int count = 0;
        
        foreach (string word in babbling)
        {
            string tmp = word;
            tmp = tmp.Replace("aya", "1");
            tmp = tmp.Replace("ye", "2");
            tmp = tmp.Replace("woo", "3");
            tmp = tmp.Replace("ma", "4");
            
            // 조건1: 치환하고 남은 글자가 없어야 함 (4개 음절로만 이루어짐)
            bool onlyValidSyllables = true;
            foreach (char c in tmp)
            {
                if (c != '1' && c != '2' && c != '3' && c != '4')
                {
                    onlyValidSyllables = false;
                    break;
                }
            }
            
            // 조건2: 같은 숫자(음절)가 연속되면 안 됨
            bool hasConsecutive = false;
            for (int i = 0; i < tmp.Length - 1; i++)
            {
                if (tmp[i] == tmp[i + 1])
                {
                    hasConsecutive = true;
                    break;
                }
            }
            
            if (onlyValidSyllables && !hasConsecutive)
            {
                count++;
            }
        }
        
        return count;
    }
}

File: 프로그래머스/1/133499. 옹알이 （2）/옹알이 （2）.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(vector<string> babbling) 
{
    map<string, char> babbMap;
    
    // map을 만드는 과정
    vector<string> key = {"aya", "ye", "woo", "ma"};
    vector<char> value = {'1', '2', '3', '4'};
    int count = 0;
    for(int i=0; i<key.size(); i++)
    {
        babbMap[key[i]] = value[i]; // ["aya"] = '1' , ["ye"] = '2' ....
    }
    
    for(string &babb : babbling)
    {
        for(int i=0; i<babbMap.size(); i++) // ["aya"] ["ye"] ["woo"] ["ma"] 순으로 체크
        {
            // key[i] => (["aya"] ["ye"] ["woo"] ["ma"])가 babb 에 포함되어있으면..
            while(babb.find(key[i]) != string::npos) 
            {
                size_t pos = babb.find(key[i]); // key[i]가 발견된 시점의 인덱스를 pos로 받아와서
                // babb의 pos시점에서부터 key[i] 만큼의 길이를 value[i]로 바꾼다.
                // 이때 value[i]는 char형이므로 3번째에 1이라는 인자를 넣어준다.
                babb.replace(pos, key[i].length(), 1, value[i]); 
            }
        }
        // 디버깅용
        cout << babb << endl;
        
        bool isOnlyNum = true; // babb이 오직 숫자로만 이루어져있는지 체크하는 변수
        for(char &i : babb)
        {
            if(!isdigit(i))
            {
                isOnlyNum = false;
                break;
            }
        }
        bool isBool = false;
        if(isOnlyNum)
        {
            for(int i=0; i<babb.length()-1; i++)
            {
                if(babb[i] == babb[i+1])
                {
                    isBool = true;
                    break;
                }
            }
        }
        if(isOnlyNum && !isBool)
        {
            count++;
        }
    }
    return count;
}